File: src/main/java/com/aquatrack/service/HouseholdService.java
package com.aquatrack.service;

import com.aquatrack.dto.household.HouseholdRequest;
import com.aquatrack.dto.household.HouseholdResponse;

import java.util.List;

/**
 * Service interface for Household Management.
 */
public interface HouseholdService {

    // ==========================================
    // Create Household
    // ==========================================

    /**
     * Creates a new household under the specified floor.
     *
     * @param floorId Floor ID
     * @param request Household request
     * @return Created household
     */
    HouseholdResponse createHousehold(
            Long floorId,
            HouseholdRequest request
    );

    // ==========================================
    // Get Households
    // ==========================================

    /**
     * Returns all households for a floor.
     *
     * @param floorId Floor ID
     * @return List of households
     */
    List<HouseholdResponse> getHouseholdsByFloor(
            Long floorId
    );

    /**
     * Returns a household by ID.
     *
     * @param floorId Floor ID
     * @param householdId Household ID
     * @return Household details
     */
    HouseholdResponse getHouseholdById(
            Long floorId,
            Long householdId
    );

    /**
     * Returns all households belonging to the buildings
     * assigned to the currently logged-in manager.
     *
     * @return List of households
     */
    List<HouseholdResponse> getManagerHouseholds();

    // ==========================================
    // Update Household
    // ==========================================

    /**
     * Updates an existing household.
     *
     * @param floorId Floor ID
     * @param householdId Household ID
     * @param request Updated household request
     * @return Updated household
     */
    HouseholdResponse updateHousehold(
            Long floorId,
            Long householdId,
            HouseholdRequest request
    );

    // ==========================================
    // Delete Household
    // ==========================================

    /**
     * Deletes a household.
     *
     * @param floorId Floor ID
     * @param householdId Household ID
     */
    void deleteHousehold(
            Long floorId,
            Long householdId
    );

}

File: Aquatrack-frontend/src/services/householdService.js
import api from "../api/axios";

const BASE_URL = "/manager";

const householdService = {

    // ==========================================
    // Dashboard
    // ==========================================

    getManagerHouseholds: async () => {

        const response = await api.get(
            `${BASE_URL}/households`
        );

        return response.data.data;

    },

    // ==========================================
    // Buildings
    // ==========================================

    getManagerBuildings: async () => {

        const response = await api.get(
            `${BASE_URL}/buildings`
        );

        return response.data?.data ?? response.data;

    },

    // ==========================================
    // Floors
    // ==========================================

    getManagerFloors: async (buildingId) => {

        const response = await api.get(
            `${BASE_URL}/buildings/${buildingId}/floors`
        );

        return response.data;

    },

    // ==========================================
    // Households by Floor
    // ==========================================

    getHouseholdsByFloor: async (floorId) => {

        const response = await api.get(
            `${BASE_URL}/floors/${floorId}/households`
        );

        return response.data.data;

    },

    getHouseholdById: async (floorId, householdId) => {

        const response = await api.get(
            `${BASE_URL}/floors/${floorId}/households/${householdId}`
        );

        return response.data.data;

    },

    createHousehold: async (floorId, data) => {

        const response = await api.post(
            `${BASE_URL}/floors/${floorId}/households`,
            data
        );

        return response.data.data;

    },

    updateHousehold: async (
        floorId,
        householdId,
        data
    ) => {

        const response = await api.put(
            `${BASE_URL}/floors/${floorId}/households/${householdId}`,
            data
        );

        return response.data.data;

    },

    deleteHousehold: async (
        floorId,
        householdId
    ) => {

        const response = await api.delete(
            `${BASE_URL}/floors/${floorId}/households/${householdId}`
        );

        return response.data;

    }

};

export default householdService;